src/coreclr/jit/objectalloc.cpp

```diff
@@ -1374,12 +1418,33 @@ void ObjectAllocator::RewriteUses()
                     }
                 }
 
-                if (lclVarDsc->lvType != newType)
+                // For local structs, retype the GC fields.
+                //
+                if (lclVarDsc->lvType == TYP_STRUCT)
+                {
+                    // We should only see spans here.
+                    //
+                    assert(lclVarDsc->IsSpan());
+                    ClassLayout* const layout    = lclVarDsc->GetLayout();
+                    ClassLayout*       newLayout = nullptr;
+
+                    if (newType == TYP_I_IMPL)
+                    {
+                        // No GC refs remain, so now a block layout
+                        newLayout = m_compiler->typGetBlkLayout(layout->GetSize());
+                        JITDUMP("Changing layout of span V%02u to block\n", lclNum);
+                        lclVarDsc->ChangeLayout(newLayout);
```

Comment:
I think for this we want a layout with proper padding information, otherwise the promotion will be suboptimal compared to the original layout.

Reply:
Ah, ok... maybe an "erase GC" method on a layout?

Reply:
> Ah, ok... maybe an "erase GC" method on a layout?

I think you can do something like:
```cpp
ClassLayoutBuilder builder(lcl->GetLayout()->GetSize());
builder.CopyInfoFrom(0, lcl->GetLayout(), /* copyPadding */ true);
builder.SetGCPtrType(OFFSETOF__CORINFO_Span__reference / TARGET_POINTER_SIZE, TYP_I_IMPL);
```

Reply:
Although I suppose if we generalized this PR to arbitrary structs we would indeed want something that only copies the padding information, so maybe introducing another parameter to `CopyInfoFrom` to ignore the GC information would be the way to go.